app/models/backup_code_configuration.rb

```diff
@@ -36,7 +39,29 @@ def find_with_code(code:, user_id:)
       return if code.blank?
       code = code.downcase.strip
       code_fingerprint = create_fingerprint(code)
-      find_by(code_fingerprint: code_fingerprint, user_id: user_id)
+
+      user_salt_costs = select(:code_salt, :code_cost).
+        distinct.
+        where(user_id: user_id).
+        where.not(code_salt: nil, code_cost: nil).
+        limit(BackupCodeGenerator::NUMBER_OF_CODES * 2).
```

Comment:
this is a very expensive query if a user somehow has many codes, so the limit is to make sure we don't trash the DB. I'm open to a bigger multiple than 2, it seemed like a safe guess -- gonna spot check the prod DB too

Reply:
You could key off of this index ["user_id", "created_at"] by including a sort by since new ones are only generated when all the old ones are used or we can just add another index ["user_id", "used_at"],

Reply:
The sort shouldn't be necessary, a `WHERE user_id = ?` clause is enough to use the index on ["user_id", "created_at"] ([docs](https://www.postgresql.org/docs/12/indexes-multicolumn.html))

>A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

I don't think we can put a limit on the query or we may deny a valid backup code. Old codes are deleted when a new set is generated, so the most any user should have is 10, which should be fine to query.

The salt will be unique every time, so we could probably drop the distinct.

Could we avoid hashing used keys?

Reply:
I scanned prod and a few users have up to 20 but nobody has more, so I think a limit of 2x is fine, but 3x would also be safe if we were worried

```ruby
User.select(:id).find_in_batches(batch_size: 100_000) do |batch|
  puts BackupCodeConfiguration.
    select('COUNT(*) AS c, user_id').
    group(:user_id).
    where(user_id: batch.map(&:id)).
    having('COUNT(*) > 10').
    map(&:c).to_a.inspect
end
# ...
[]
[17]
[]
[19]
[20]
[11, 16, 16, 19, 12, 18]
[12, 15, 20, 18]
# ...
```

Reply:
I'm curious how that happened, but I think we still have to keep it unlimited for backwards compatibility. Selecting < 100 rows that are well-indexed shouldn't be too much of a burden on the DB?

Reply:
If you have a query that is not fully covered in the index, even if you have the compound index [user_id,created_at]  where logically we know it should seek to the first user_id record and grab the rest using the selected criteria like code_salt and code_cost, the optimizer might decide instead to do a db full scan and the db is toast.  This is because the user can have potentially millions of codes that the selection criteria would need to sift through.  It uses db statistics to make a best guess but it doesn't always guess right.

Reply:
>This is because the user can have potentially millions of codes that the selection criteria would need to sift through

Per above, the most a given user has is 20. Does that change the calculation? The query would also not be fully covered unless code_salt and code_cost were also indexed. Postgres 11 added the ability to do a covering index, but that's not an option yet 🙂

Reply:
All this talk on indexes reminded me to add an index for the new column so we can drop the old one 30650e1

Reply:
Haha. Great!  wrt changing calc and "millions of codes" it def increases the likelihood that all is well since we limit the codes for a user.  If I had to bet I'd say we are ok but we can just check the plan by the optimizer beforehand.  A similar situation where our optimizer made the wrong decision was `take` vs `first` in phone_configurations.  Generally a user has very few phones.  Technically `id` is not covered in our indexes on a query by `user_id` but that is what `first` uses (lowest id).  So instead of grabbing all the user's phones by user_id and sorting them the optimizer proceeded to do a full scan by id.  That's why I usually add a compound index with `created_at` if we need the first x or pagination.